src/GafferArnold/IECoreArnoldPreview/Renderer.cpp

```diff
@@ -3465,45 +3465,80 @@ class ArnoldRenderer final : public ArnoldRendererBase
 
 		ArnoldRenderer( RenderType renderType, const std::string &fileName, const IECore::MessageHandlerPtr &messageHandler )
 			:	ArnoldRendererBase( nodeDeleter( renderType ) ),
-				m_globals( new ArnoldGlobals( renderType, fileName, m_shaderCache.get() ) )
+				m_messageHandler( messageHandler ), m_globals( new ArnoldGlobals( renderType, fileName, m_shaderCache.get() ) )
 		{
+			if( m_messageHandler )
+			{
+				AiMsgSetCallback( &ArnoldRenderer::aiMsgCallback );
+			}
 		}
 
 		~ArnoldRenderer() override
 		{
 			pause();
+			if( g_currentMessageHandler == m_messageHandler )
+			{
+				g_currentMessageHandler = nullptr;
+			}
 		}
 
 		void option( const IECore::InternedString &name, const IECore::Object *value ) override
 		{
+			const IECore::MessageHandler::Scope s( m_messageHandler.get() );
 			m_globals->option( name, value );
 		}
 
 		void output( const IECore::InternedString &name, const IECoreScene::Output *output ) override
 		{
+			const IECore::MessageHandler::Scope s( m_messageHandler.get() );
 			m_globals->output( name, output );
 		}
 
 		ObjectInterfacePtr camera( const std::string &name, const IECoreScene::Camera *camera, const AttributesInterface *attributes ) override
 		{
+			const IECore::MessageHandler::Scope s( m_messageHandler.get() );
 			m_globals->camera( name, camera );
 			return ArnoldRendererBase::camera( name, camera, attributes );
 		}
 
 		void render() override
 		{
+			const IECore::MessageHandler::Scope s( m_messageHandler.get() );
+			// Promote our handler to global
+			if( m_messageHandler )
+			{
+				g_currentMessageHandler = m_messageHandler;
+			}
```

Comment:
Why do this here rather than in the constructor? The symmetry of setting up in the constructor and tearing down in the destructor has a certain logical appeal.

Reply:
I was thinking of the situation we have where shader balls and an IPR node may both have made a renderer. Let me tooth-comb the implications of exactly what happens.  I think I wanted to make sure that re-starting an IPR node would re-instate the correct handler. But I think we might be making a new renderer then anyway.

Reply:
I would expect the ArnoldRenderer constructor to throw if there is any other ArnoldRenderer instance alive, regardless of paused status.

Reply:
Arnold support refactored in https://github.com/GafferHQ/gaffer/pull/3767/commits/d917f02a78e2c4a5bc1f11efa98d8fed04f9c996